messagebus/src/vespa/messagebus/network/rpcsendv2.cpp

```diff
@@ -0,0 +1,247 @@
+    root.setLong(RETRY_F, msg.getRetry());
+    root.setLong(TIMELEFT_F, timeRemaining);
+    root.setString(PROTOCOL_F, msg.getProtocol());
+    root.setLong(TRACELEVEL_F, traceLevel);
+    filler.fill(BLOB_F, root);
+
+    OutputBuf rBuf(8192);
+    BinaryFormat::encode(slime, rBuf);
+    ConstBufferRef toCompress(rBuf.getBuf().getData(), rBuf.getBuf().getDataLen());
+    DataBuffer buf(vespalib::roundUp2inN(rBuf.getBuf().getDataLen()));
+    CompressionConfig::Type type = compress(_net->getCompressionConfig(), toCompress, buf, false);
+
+    FRT_Values &args = *req.GetParams();
+    req.SetMethodName(METHOD_NAME);
+    args.AddInt8(type);
+    args.AddInt32(toCompress.size());
+    args.AddData(buf.stealBuffer(), buf.getDataLen());
+}
+
+namespace {
+
+class ParamsV2 : public RPCSend::Params
+{
+public:
+    ParamsV2(const FRT_Values &arg)
+        : _slime()
+    {
+        uint8_t encoding = arg[0]._intval8;
+        uint32_t uncompressedSize = arg[1]._intval32;
+        DataBuffer uncompressed(arg[2]._data._buf, arg[2]._data._len);
+        ConstBufferRef blob(arg[2]._data._buf, arg[2]._data._len);
+        decompress(CompressionConfig::toType(encoding), uncompressedSize, blob, uncompressed, true);
+        assert(uncompressedSize == uncompressed.getDataLen());
+        BinaryFormat::decode(Memory(uncompressed.getData(), uncompressed.getDataLen()), _slime);
+    }
+
+    uint32_t getTraceLevel() const override { return _slime.get()[TRACELEVEL_F].asLong(); }
+    bool useRetry() const override { return _slime.get()[USERETRY_F].asBool(); }
+    uint32_t getRetries() const override { return _slime.get()[RETRY_F].asLong(); }
+    uint64_t getRemainingTime() const override { return _slime.get()[TIMELEFT_F].asLong(); }
+
+    Version getVersion() const override {
+        return Version(_slime.get()[VERSION_F].asString().make_stringref());
+    }
+    stringref getRoute() const override {
+        return _slime.get()[ROUTE_F].asString().make_stringref();
+    }
+    stringref getSession() const override {
+        return _slime.get()[SESSION_F].asString().make_stringref();
+    }
+    stringref getProtocol() const override {
+        return _slime.get()[PROTOCOL_F].asString().make_stringref();
+    }
+    BlobRef getPayload() const override {
+        Memory m = _slime.get()[BLOB_F].asData();
+        return BlobRef(m.data, m.size);
+    }
+private:
+    Slime _slime;
+};
+
+}
+
+std::unique_ptr<RPCSend::Params>
+RPCSendV2::toParams(const FRT_Values &args) const
+{
+    return std::make_unique<ParamsV2>(args);
+}
+
+std::unique_ptr<Reply>
+RPCSendV2::createReply(const FRT_Values & ret, const string & serviceName,
+                       Error & error, vespalib::TraceNode & rootTrace) const
+{
+    uint8_t encoding = ret[0]._intval8;
+    uint32_t uncompressedSize = ret[1]._intval32;
+    DataBuffer uncompressed(ret[2]._data._buf, ret[2]._data._len);
+    ConstBufferRef blob(ret[2]._data._buf, ret[2]._data._len);
```

Comment:
Dangerous to trust buffer lengths received from the network. Should check lengths against received blob size.

Reply:
Ignore comment, did not originally see that length is from RPC layer itself 🦄 👍